### ISSUE TITLE
Distribution was not found and is required by the application

### ISSUE THREAD

**Issue by Samuel-Nathanson** [pyinstaller/pyinstaller#4409]
Hi,

I have opencv matrices in C++ that I want to process with Python, but I don't want to install Python on the end-users machine, and I don't want to have interprocess communication between C++ and my Python code. I am looking to decide between Nuitka, CPython, PyOxidizer, and PyInsatller. 

Currently, with PyInstaller, I can build the standalone application with 
`pyinstaller CSR_loadAndInferOneImage.py`
However, I get a run-time error during the execution of the application. 
pkg_resources.DistributionNotFound: The 'fastprogress>=0.1.19' distribution was not found and is required by the application
[19757] Failed to execute script CSR_loadAndInferOneImage

When I run the script in python, I do not get this error

In the Python REPL, I can run 
>>> import fastprogress
>>> fastprogress.__version__
'0.1.21'

Why could this be occurring?

**Comment by LightCC** [comment]
I'm having a similar problem, and from searching on StackOverflow and other places, it appears this is a common issue when dealing with version or other package metadata.

I'm not 100% sure that the package metadata is always stored in the 'my_project.egg-info' folder, but if you have such a folder in the fastprogress package, then you will need to include it in your *.spec file under the 'datas=' line.

I was able to get mine working by updating my `myscript.spec` file `a =` lines as follows, with the `datas=` manually added:

```python
a = Analysis(['myscript.py'],
             pathex=['C:\\path\\to\\mypackage'],
             binaries=[],
             datas=[('mypackage.egg-info/*','mypackage.egg-info')],
             hiddenimports=[],
             hookspath=[],
             runtime_hooks=[],
             excludes=[],
             win_no_prefer_redirects=False,
             win_private_assemblies=False,
             cipher=block_cipher,
             noarchive=False)
```

However, this appears to be one of the most common failure modes, based on stackoverflow questions and the way the failure manifests. I was hoping for a more elegant solution. It seems pyinstaller should be able to detect that pkg_resources or distutils is being accessed and decide to add the module meta-data, or have some way of annotating that it is needed in the source scripts.

One of the answers updated the hook files with the following. If there is a way to store a hook file for your local package in the package itself (so it can be captured with version control), this would be a much better solution:

```python
from PyInstaller.utils.hooks import copy_metadata

datas = copy_metadata('md2mat')
```

**Comment by htgoebel** [comment]
@Samuel-Nathanson: @LightCC pointed to the solution: implementing a hook.

We would appreciate if you'd submit a pull-request for a hook for this. See [this section of the manual](https://pyinstaller.readthedocs.org/en/latest/hooks.html) for more information about how to implement hooks. See our Development Guide on [how to create pull-requests for PyInstaller](https://pyinstaller.readthedocs.io/en/stable/development/pull-requests.html). Thanks.

**Comment by LightCC** [comment]
For the record, I did figure out how to store the hooks file in the repo (so it is saved with my VCS), and include it with pyinstaller at the command-line, as follows.

It would still be better if pyinstaller could auto-detect the need for the metadata (perhaps this can be linked in with the new built-in `importlib-metadata` import now available in 3.8.0), but here's how to do it at the moment:

Create a hook file - `hook-module.py`, (substitute one of your custom module names for `module`, ) anywhere you want in your package. This must be a module that gets imported in your scripts. I just used the module name that is the same as my package name.

I have the semi-standard package setup with a source folder, a tests folder, etc. and I chose to add a `hooks` folder and placed my hook file there.  Put the following in the hook file:

```py
from PyInstaller.utils.hooks import copy_metadata

datas = copy_metadata('mypackage')
```

where you substitute your package name for `mypackage` (with quotes around it, as a string). Note that if you define or update the `datas` variable somewhere else for PyInstaller, you will need to modify the above to account for that.

Then when running PyInstaller, just add the `--additional-hooks-dir=.\\hooks` argument. This assumes you added a `hooks` directory to your package root folder, and with a Windows style path.

That's it. PyInstaller will search the additional hooks folder, and find the custom hook file and use it. 

_**Note:** You still have to make sure that the metadata is there and has been updated before running PyInstaller. For example, you might need to run `python setup.py sdist` to get setuptools to update the metadata, or may need to update it directly, depending how you have your package setup._
